Marlin/src/core/serial_base.h

```diff
@@ -82,9 +82,18 @@ struct SerialBase {
   NO_INLINE void print(unsigned char c, int base = 0)     { print((unsigned long)c, base); }
   NO_INLINE void print(int c, int base = DEC)             { print((long)c, base); }
   NO_INLINE void print(unsigned int c, int base = DEC)    { print((unsigned long)c, base); }
-  void print(long c, int base = DEC)                      { if (!base) write(c); else { write((const uint8_t*)"-", c < 0); printNumber(c < 0 ? -c : c, base); } }
   void print(unsigned long c, int base = DEC)             { printNumber(c, base); }
   void print(double c, int digits = 2)                    { printFloat(c, digits); }
+  void print(long c, int base = DEC)                      {
+    if (!base) {
+      write(c);
+      return;
+    }
+    if (base == DEC && c < 0) {
+      write((uint8_t)'-'); c = -c;
```

Comment:
write('-') ;is more than enough

Reply:
ar at worst (char)'-'

Reply:
Yep probably but on x86, it'll trigger a "conversion from signed to unsigned" warning if not cast.

Reply:
BTW, there is a latent bug in `c = -c` for `LONG_MIN` which does not exist (since `-LONG_MIN > LONG_MAX`) and can not be represented in a signed long... This was present in Marlin previously, so I don't know if I should fix it or not.

Reply:
This only happens for platform without 2-complement representation for negative number. I think all platform are 2-complement so it's ok.

Reply:
it should be wrong then where it0s already used a char in other part of the code (G29 T and other parts)

Reply:
Wait, isn't that backwards? 2's-complement numbers _do_ have the problem. If someone calls `print((uint16_t)0x8000)`, then it'll try to compute a 16-bit equivalent of -(-32768), which doesn't exist.

Reply:
Er, well, it's `long`, so substitute `(uint32_t)0x80000000` and `-(-2147483648)`.

Reply:
I just copy/pasted the new implementation to a stand-alone test project running on my x86 desktop, with `printNumber` implemented as simply `cout << n`, and when I passed in 0x80000000, the output it generated was `--2147483648`.

Reply:
Haha, I just also copy/pasted in the definition of `printNumber`, and it prints `0x80000000` as `-56303/413/`. :-D

Reply:
Here's one possible mitigation which works on my desktop and doesn't seem to depend on a particular binary representation of the number, handling negatives as check one layer down in `printNumber` instead:
```
void printNumber(unsigned long n, const uint8_t base)
{
  if (!base) {
    write((uint8_t)n);
    return;
  }
  if (n) {
    unsigned char buf[8 * sizeof(long)]; // Enough space for base 2
    int8_t i = 0;
    signed long sn = (signed long)n;
    if ((sn < 0) && (base == DEC)) {
      write((uint8_t)'-');
      // Handle the first digit specially. In an earlier version of this code, we would simply set n to -n, but
      // there might exist a value of n for which we can't represent -n directly.
      buf[i++] = -(sn % base);
      n = (unsigned long)(sn / -base);
    }
    while (n) {
      buf[i++] = n % base;
      n /= base;
    }
    while (i--) write((char)(buf[i] + (buf[i] < 10 ? '0' : 'A' - 10)));
  }
  else write('0');
}
```
With this change, the handling of negatives inside `print` is no longer needed.

Testing:
```
  for (int base = 2; base <= 16; base++)
  {
    print(0x80000000, base);
    write('\n');
  }
```

```
10000000000000000000000000000000
12112122212110202102
2000000000000000
13344223434043
553032005532
104134211162
20000000000
5478773672
-2147483648
A02220282
4BB2308A8
282BA4AAB
1652CA932
C87E66B8
80000000
```

I have confirmed these values. I also tested the sequence 0x7FFFFFF0 through 0x8000000F in each of these bases and didn't see any anomalies.

Reply:
@logiclrd Thanks for the example code. Indeed, the initial serial code print had issues. The new code still have them as well (I've copy and pasted here the old code). You're right but IMHO, it's a 2 step process. First, we make a single version of the print function in all the code base, then we fix it. Your code will happen in a second step, for now, I'd like to keep bug-for-bug equivalent code... if I can say so, to ensure I'm not introducing too many differences in one shot.

Reply:
Please let me know your opinion.

Reply:
Ideally, I'd like to remove all special case print with base == 0, since when you do call print with base = 0, you should have called write instead (it's only for char and unsigned char). This means replacing in the code base all call to `SERIAL_ECHO(uint8)` when what was really meant is `SERIAL_CHAR()`. That way, once you read in the code a SERIAL_ECHO, it **always** mean printing a number as decimal (or at worst, in another base), but not outputing an ASCII char.

Reply:
Once this is done, we'll be able to simplify the print functions not to have so many overload/case. In your code, you're computing a modulo and that's expensive (although, the only base we are interested in is 10 and 16 in reality). If you compute modulo 16 **and** you have 2 complement number, then the modulo is not required.

Reply:
I'm

Reply:
I'm not convinced that that level of optimization is really necessary. What proportion of the available CPU cycles is actually used by formatting integers? We could make it blindingly fast with a separate path per base and careful bit arithmetic, but is that solving any problem?

I'm only a bystander, a user of the software and not a developer of it, but I am a developer professionally and can say wholeheartedly that I am 100% on board with refactoring and reworking as entirely independent steps. :-)

I might say that if different parts of the code have different interpretations of `SERIAL_ECHO` then perhaps they should _all_ have to change -- make `SERIAL_WRITE_NUMBER` and `SERIAL_WRITE_CHAR` and get rid of `SERIAL_ECHO` entirely, and every call site changes to a macro that is truly explicit about what it is doing.

Reply:
Yep. I'm inline with your opinion.

Reply:
Our assumption has always been that an 8-bit number (`byte`, `char`, `uint8_t`, etc.) *may* be printed as a single character by the serial methods, so you will see many places where an 8-bit value is cast to an integer in order to get the desired numerical output.